ui/raidboss/popup-text.ts

```diff
@@ -978,113 +1204,174 @@ export class PopupText {
-        const duration = triggerHelper.duration.fromConfig ||
-          triggerHelper.duration.fromTrigger || triggerHelper.duration[lowerTextKey];
+        let duration = triggerHelper.duration?.fromConfig ||
+          triggerHelper.duration?.fromTrigger;
+        if (duration === undefined && triggerHelper.duration)
+          duration = triggerHelper.duration[lowerTextKey];
+        if (duration === undefined)
+          duration = 0;
+
         this._createTextFor(triggerHelper, text, textType, lowerTextKey, duration);
         if (!triggerHelper.soundUrl) {
-          triggerHelper.soundUrl = this.options[textTypeUpper + 'Sound'];
-          triggerHelper.soundVol = this.options[textTypeUpper + 'SoundVolume'];
+          triggerHelper.soundUrl = this.options[upperSound];
+          triggerHelper.soundVol = this.options[upperSoundVolume];
         }
       }
     }
   }
 
-  _makeTextElement(triggerHelper, text, className) {
+  _makeTextElement(_triggerHelper: TriggerHelper, text: string, className: string): HTMLElement {
     const div = document.createElement('div');
     div.classList.add(className);
     div.classList.add('animate-text');
     div.innerText = text;
     return div;
   }
 
-  _playAudioFile(triggerHelper, url, volume) {
+  _playAudioFile(triggerHelper: TriggerHelper, url: string, volume: number): void {
     const audio = new Audio(url);
     audio.volume = volume || 1;
-    audio.play();
+    void audio.play();
   }
 
-  Test(zone, log) {
-    this.OnPlayerChange({ detail: { name: 'ME' } });
-    this.OnChangeZone({ zoneName: zone });
-    this.OnLog({ detail: { logs: ['abcdefgh', log, 'hgfedcba'] } });
+  getDataObject(): RaidbossData {
+    let preserveHP = 0;
+    if (this.data && this.data.currentHP)
+      preserveHP = this.data.currentHP;
+
+    // TODO: make a breaking change at some point and
+    // make all this style consistent, sorry.
+    return {
+      me: this.me,
+      job: this.job,
+      role: this.role,
+      party: this.partyTracker,
+      lang: this.parserLang,
+      parserLang: this.parserLang,
+      displayLang: this.displayLang,
+      currentHP: preserveHP,
+      options: this.options,
+      ShortName: this.ShortNamify.bind(this),
+      StopCombat: () => this.SetInCombat(false),
+      ParseLocaleFloat: parseFloat,
+      CanStun: () => Util.canStun(this.job),
+      CanSilence: () => Util.canSilence(this.job),
+      CanSleep: () => Util.canSleep(this.job),
+      CanCleanse: () => Util.canCleanse(this.job),
+      CanFeint: () => Util.canFeint(this.job),
+      CanAddle: () => Util.canAddle(this.job),
+    };
   }
 }
 
 export class PopupTextGenerator {
-  constructor(popupText) {
-    this.popupText = popupText;
+  constructor(private popupText: PopupText) {
   }
 
-  Info(text) {
+  Info(text: string): void {
     this.popupText.OnTrigger({
       infoText: text,
       tts: text,
-    });
+    },
+    null,
+    Date.now());
```

Comment:
Should OnTrigger do:
```ts
  this.OnTriggerInternal(trigger, matches, currentTime ?? Date.now());
```
And then matches and currentTime could be optional

Reply:
`currentTime` is deliberately not optional for emulator purposes, so I don't think this is a good idea. Realistically, timeline should be passing a correct DateTime to these helpers as it is but that's beyond the scope of this PR.

To be more specific, `timeline` should be passing `this.timebase` through. The goal is to never initiate a timestamp directly from current time, and instead always have it derived from the log line's timestamp.